[3/n] [c10d] Revise the implementation of TCPStore

cbalioglu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #58331 [4/n] [c10d] Introduce the multi-tenancy feature in TCPStore
* **#58330 [3/n] [c10d] Revise the implementation of TCPStore**
* #58329 [2/n] [c10d] Introduce the 'multiTenant' constructor parameter in TCPStore
* #58328 [1/n] [c10d] Introduce a new TCPStore constructor

This PR is part of a stack that addresses the GitHub issue #41614; it introduces a major refactoring of the `TCPStore` class in preparation of the multi-tenancy feature.

- All TCP sockets are wrapped with a new `TCPSocket` RAII type.
- `BackgroundThread` and daemon types are moved from header to cpp file.
- Server, client, and callback sockets are refactored into their own internal types `TCPServer`, `TCPClient` and `TCPCallbackClient`.
- Calls to `tcputil::send*` and `tcputil::recv*` are wrapped in `TCPClient` for easier readability and maintenance purposes.
- Two `TODO` statements are put to reference future improvements. Based on feedback, I will either create separate GitHub issues for them or address them as part of this stack.

Differential Revision: [D28448981](https://our.internmc.facebook.com/intern/diff/D28448981/)

cbalioglu committed:
[3/n] [c10d] Revise the implementation of TCPStore

This PR is part of a stack that addresses the GitHub issue #41614; it introduces a major refactoring of the `TCPStore` class in preparation of the multi-tenancy feature.

- All TCP sockets are wrapped with a new `TCPSocket` RAII type.
- Server, client, and callback sockets are refactored into their own internal types `TCPServer`, `TCPClient` and `TCPCallbackClient`.
- Calls to `tcputil::send*` and `tcputil::recv*` are wrapped in `TCPClient` for easier readability and maintenance purposes.
- Two `TODO` statements are put to reference future improvements. Based on feedback, I will either create separate GitHub issues for them or address them as part of this stack.

Differential Revision: [D28448981](https://our.internmc.facebook.com/intern/diff/D28448981/)

[ghstack-poisoned]

cbalioglu committed:
Update on "[3/n] [c10d] Revise the implementation of TCPStore"


This PR is part of a stack that addresses the GitHub issue #41614; it introduces a major refactoring of the `TCPStore` class in preparation of the multi-tenancy feature.

- All TCP sockets are wrapped with a new `TCPSocket` RAII type.
- Server, client, and callback sockets are refactored into their own internal types `TCPServer`, `TCPClient` and `TCPCallbackClient`.
- Calls to `tcputil::send*` and `tcputil::recv*` are wrapped in `TCPClient` for easier readability and maintenance purposes.
- Two `TODO` statements are put to reference future improvements. Based on feedback, I will either create separate GitHub issues for them or address them as part of this stack.

Differential Revision: [D28448981](https://our.internmc.facebook.com/intern/diff/D28448981/)

[ghstack-poisoned]

mrshenli commented on torch/lib/c10d/TCPStore.cpp:
```diff
@@ -700,92 +888,74 @@ void TCPStore::waitForWorkers() {
-
-  // Block until callback has been registered successfully
-  tcpStoreWorkerDaemon_->waitForCallbackRegistration();
+  callback_client_->setCallback(keyPrefix_ + key, callback);
```
@H-Huang do the changes on watchKey and `TCPStoreMasterDaemon` / `TCPStoreWorkerDaemon` classes look OK to you?

H-Huang replied:
watchKey refactoring looks okay, will leave feedback for TCPStoreMasterDaemon / TCPStoreWorkerDaemon in separate comment

mrshenli commented on torch/lib/c10d/TCPStore.hpp:
```diff
@@ -125,6 +125,21 @@ class TCPStoreWorkerDaemon : public BackgroundThread {
   bool callbackRegisteredData_ = false;
 };
 
+namespace detail {
```
curious about this detail namespace. Is it because we don't want to pollute the parent c10d namespace as there might be future applications adopting the c10d extension APIs?

cbalioglu replied:
`TCPServer`, `TCPClient` and `TCPCallbackClient` are implementation details of `TCPStore`. Since they have to be member variables of `TCPStore` I forward declared them in the public header file and used pimpl idiom to hide their definitions. I also deliberately put them in the `detail` namespace to clearly indicate that they are not part of the public API.

mrshenli commented on torch/lib/c10d/TCPStore.cpp:
```diff
@@ -591,6 +591,229 @@ void TCPStoreMasterDaemon::run() {
+      : socket_{std::move(socket)}, port_{port}, daemon_{std::move(daemon)} {}
+
+ private:
+  TCPSocket socket_;
+  PortType port_;
```
can these be const?

cbalioglu replied:
Technically yes, however, I usually avoid using `const` qualifier for member variables as they interfere with move semantics (compiler silently fallbacks to copy semantic due to "constness" of a variable). In this particular case it should be safe though as we always heap allocate `TCPServer` and move it around with a `std::unique_ptr<>`. I will make them `const` in the next revision.

mrshenli commented on torch/lib/c10d/TCPStore.cpp:
```diff
@@ -591,6 +591,229 @@ void TCPStoreMasterDaemon::run() {
+
+  void sendCommandForKey(QueryType type, const std::string& key);
+
+  void sendBits(const std::vector<std::uint8_t>& value) {
```
Bits -> Bytes?

cbalioglu replied:
Sounds fair :) will rename it.

mrshenli commented on torch/lib/c10d/TCPStore.cpp:
```diff
@@ -591,6 +591,229 @@ void TCPStoreMasterDaemon::run() {
+    tcputil::sendVector<std::uint8_t>(socket_.handle(), value);
+  }
+
+  void sendStrings(c10::ArrayRef<std::string> value);
```
curious, any reason for using `ArrayRef` for string but `const std::vector&` for bytes?

cbalioglu replied:
For bytes I am forwarding the call to `tcputil::sendVector()` which expects a `const std::vector&`. I just did not want to touch `tcputil` in this particular task since it could affect other parts of the code base. If you think that it is relatively safe, I can revise the parameter type of `tcputil::sendVector()`.

mrshenli commented on torch/lib/c10d/TCPStore.cpp:
```diff
@@ -591,6 +591,229 @@ void TCPStoreMasterDaemon::run() {
+      sizeof(timeoutTV)));
+}
+
+class TCPCallbackClient {
```
Is there any reason for creating a dedicated `TCPCallbackClient` class, instead making this as a feature on TCPClient?

cc @H-Huang

cbalioglu replied:
I refactored the code in a way that each socket is abstracted into its own type (`TCPServer`, `TCPClient`, `TCPCallbackClient`) to make sure I maintain single responsibility principle for each type. Since the callback mechanism uses its own socket, I ended up with `TCPCallbackClient`.

H-Huang replied:
This reasoning makes sense to me

mrshenli reviewed (commented):
I might not be the best person to review the changes in TCPStore. cc @H-Huang  and @osalpekar for additional reviews.

H-Huang commented on torch/lib/c10d/TCPStore.cpp:
```diff
@@ -591,6 +591,229 @@ void TCPStoreMasterDaemon::run() {
+  TCPSocket socket{};
+  PortType port{};
+
+  std::tie(socket, port) = tcputil::listen(opts.port);
+
+  auto daemon = std::make_unique<TCPStoreMasterDaemon>(socket.handle());
+
+  return std::make_shared<TCPServer>(
+      std::move(socket), port, std::move(daemon));
```
Previously, TCPStoreMasterDaemon was responsible for shutting down and cleaning up the socket that it used once it was done (see `BackgroundThread::~BackgroundThread()`). Now that we have this new TCPSocket class where each socket can clean itself up, I think we should also remove this part of the code:
https://github.com/pytorch/pytorch/blob/ec89bf65352e0cf1e24738cd383c94785db3d4c2/torch/lib/c10d/TCPStore.cpp#L50-L54

H-Huang reviewed (commented):
Thanks for these changes @cbalioglu, the code looks much cleaner and TCPStore constructor has been simplified a lot. The test failures for windows (https://www.internalfb.com/intern/opensource/ci/job/log/815762679367331/) do seem related to the new changes,  but unfortunately there is no helpful error message. I believe it has to do with cleanup as it aborts after the first successful TCPStore test. I left a comment regarding the socket cleanup, that may fix it.

cbalioglu committed:
Update on "[3/n] [c10d] Revise the implementation of TCPStore"


This PR is part of a stack that addresses the GitHub issue #41614; it introduces a major refactoring of the `TCPStore` class in preparation of the multi-tenancy feature.

- All TCP sockets are wrapped with a new `TCPSocket` RAII type.
- Server, client, and callback sockets are refactored into their own internal types `TCPServer`, `TCPClient` and `TCPCallbackClient`.
- Calls to `tcputil::send*` and `tcputil::recv*` are wrapped in `TCPClient` for easier readability and maintenance purposes.
- Two `TODO` statements are put to reference future improvements. Based on feedback, I will either create separate GitHub issues for them or address them as part of this stack.

Differential Revision: [D28448981](https://our.internmc.facebook.com/intern/diff/D28448981/)

[ghstack-poisoned]

cbalioglu committed:
Update on "[3/n] [c10d] Revise the implementation of TCPStore"


This PR is part of a stack that addresses the GitHub issue #41614; it introduces a major refactoring of the `TCPStore` class in preparation of the multi-tenancy feature.

- All TCP sockets are wrapped with a new `TCPSocket` RAII type.
- `BackgroundThread` and daemon types are moved from header to cpp file.
- Server, client, and callback sockets are refactored into their own internal types `TCPServer`, `TCPClient` and `TCPCallbackClient`.
- Calls to `tcputil::send*` and `tcputil::recv*` are wrapped in `TCPClient` for easier readability and maintenance purposes.
- Two `TODO` statements are put to reference future improvements. Based on feedback, I will either create separate GitHub issues for them or address them as part of this stack.

Differential Revision: [D28448981](https://our.internmc.facebook.com/intern/diff/D28448981/)

[ghstack-poisoned]

cbalioglu committed:
Update on "[3/n] [c10d] Revise the implementation of TCPStore"


This PR is part of a stack that addresses the GitHub issue #41614; it introduces a major refactoring of the `TCPStore` class in preparation of the multi-tenancy feature.

- All TCP sockets are wrapped with a new `TCPSocket` RAII type.
- `BackgroundThread` and daemon types are moved from header to cpp file.
- Server, client, and callback sockets are refactored into their own internal types `TCPServer`, `TCPClient` and `TCPCallbackClient`.
- Calls to `tcputil::send*` and `tcputil::recv*` are wrapped in `TCPClient` for easier readability and maintenance purposes.
- Two `TODO` statements are put to reference future improvements. Based on feedback, I will either create separate GitHub issues for them or address them as part of this stack.

Differential Revision: [D28448981](https://our.internmc.facebook.com/intern/diff/D28448981/)

[ghstack-poisoned]

cbalioglu commented:
> Thanks for these changes @cbalioglu, the code looks much cleaner and TCPStore constructor has been simplified a lot. The test failures for windows (https://www.internalfb.com/intern/opensource/ci/job/log/815762679367331/) do seem related to the new changes, but unfortunately there is no helpful error message. I believe it has to do with cleanup as it aborts after the first successful TCPStore test. I left a comment regarding the socket cleanup, that may fix it.

Thanks @H-Huang! I have a new revision that extends the refactoring to the daemon classes as well. I replaced all raw socket handles with `TCPSocket`. The problem with the Windows build was that our destructor for `socket_` was incorrectly called before the `daemon_` member variable in `TCPServer`. Per C++ Standard the member variables should have been destructed in reverse order, but apparently that is not the case with Visual C++.

cbalioglu committed:
Update on "[3/n] [c10d] Revise the implementation of TCPStore"


This PR is part of a stack that addresses the GitHub issue #41614; it introduces a major refactoring of the `TCPStore` class in preparation of the multi-tenancy feature.

- All TCP sockets are wrapped with a new `TCPSocket` RAII type.
- `BackgroundThread` and daemon types are moved from header to cpp file.
- Server, client, and callback sockets are refactored into their own internal types `TCPServer`, `TCPClient` and `TCPCallbackClient`.
- Calls to `tcputil::send*` and `tcputil::recv*` are wrapped in `TCPClient` for easier readability and maintenance purposes.
- Two `TODO` statements are put to reference future improvements. Based on feedback, I will either create separate GitHub issues for them or address them as part of this stack.

Differential Revision: [D28448981](https://our.internmc.facebook.com/intern/diff/D28448981/)

[ghstack-poisoned]

H-Huang reviewed (approved):
Logic changes look good to me, will also help to investigate the singular windows failure and how to resolve it, stamping to unblock

cbalioglu committed:
Update on "[3/n] [c10d] Revise the implementation of TCPStore"


This PR is part of a stack that addresses the GitHub issue #41614; it introduces a major refactoring of the `TCPStore` class in preparation of the multi-tenancy feature.

- All TCP sockets are wrapped with a new `TCPSocket` RAII type.
- `BackgroundThread` and daemon types are moved from header to cpp file.
- Server, client, and callback sockets are refactored into their own internal types `TCPServer`, `TCPClient` and `TCPCallbackClient`.
- Calls to `tcputil::send*` and `tcputil::recv*` are wrapped in `TCPClient` for easier readability and maintenance purposes.
- Two `TODO` statements are put to reference future improvements. Based on feedback, I will either create separate GitHub issues for them or address them as part of this stack.

Differential Revision: [D28448981](https://our.internmc.facebook.com/intern/diff/D28448981/)

[ghstack-poisoned]

cbalioglu committed:
Update on "[3/n] [c10d] Revise the implementation of TCPStore"


This PR is part of a stack that addresses the GitHub issue #41614; it introduces a major refactoring of the `TCPStore` class in preparation of the multi-tenancy feature.

- All TCP sockets are wrapped with a new `TCPSocket` RAII type.
- `BackgroundThread` and daemon types are moved from header to cpp file.
- Server, client, and callback sockets are refactored into their own internal types `TCPServer`, `TCPClient` and `TCPCallbackClient`.
- Calls to `tcputil::send*` and `tcputil::recv*` are wrapped in `TCPClient` for easier readability and maintenance purposes.
- Two `TODO` statements are put to reference future improvements. Based on feedback, I will either create separate GitHub issues for them or address them as part of this stack.

Differential Revision: [D28448981](https://our.internmc.facebook.com/intern/diff/D28448981/)

[ghstack-poisoned]

cbalioglu committed:
Update on "[3/n] [c10d] Revise the implementation of TCPStore"


This PR is part of a stack that addresses the GitHub issue #41614; it introduces a major refactoring of the `TCPStore` class in preparation of the multi-tenancy feature.

- All TCP sockets are wrapped with a new `TCPSocket` RAII type.
- `BackgroundThread` and daemon types are moved from header to cpp file.
- Server, client, and callback sockets are refactored into their own internal types `TCPServer`, `TCPClient` and `TCPCallbackClient`.
- Calls to `tcputil::send*` and `tcputil::recv*` are wrapped in `TCPClient` for easier readability and maintenance purposes.
- Two `TODO` statements are put to reference future improvements. Based on feedback, I will either create separate GitHub issues for them or address them as part of this stack.

Differential Revision: [D28448981](https://our.internmc.facebook.com/intern/diff/D28448981/)

[ghstack-poisoned]
